exporter/signalfxexporter/dpclient.go

```diff
@@ -48,37 +51,61 @@ type sfxDPClient struct {
 
 func (s *sfxDPClient) pushMetricsData(
 	ctx context.Context,
-	md consumerdata.MetricsData,
+	md pdata.Metrics,
 ) (droppedTimeSeries int, err error) {
-	accessToken := s.retrieveAccessToken(md)
+	metricDataByToken := s.metricsByAccessToken(md)
+	var numDroppedTimeseries int
+	var errs []error
+
+	for accessToken, metricsData := range metricDataByToken {
+		// TODO:
+		// 1) To leverage the queued retry mechanism available in new exporters,
+		// 	  look for error cases that are non-permanent and track the list of metric
+		//    data that needs to be retried. This will be done once partial retry support
+		//    is added for metrics type.
+		// 2) Also consider invoking in a goroutine to some concurrency going, in case
+		//    there are a lot of tokens.
+		droppedCount, err := s.pushMetricsDataForToken(metricsData, accessToken)
+		numDroppedTimeseries += droppedCount
+		errs = append(errs, err)
+	}
+
+	return numDroppedTimeseries, componenterror.CombineErrors(errs)
+}
 
-	sfxDataPoints, numDroppedTimeseries := s.converter.MetricDataToSignalFxV2(md)
+func (s *sfxDPClient) pushMetricsDataForToken(
+	metricsData []consumerdata.MetricsData, accessToken string) (int, error) {
+	numTimeseries := timeseriesCount(metricsData)
+	sfxDataPoints, numDroppedTimeseries := s.converter.MetricDataToSignalFxV2(metricsData)
 
 	body, compressed, err := s.encodeBody(sfxDataPoints)
 	if err != nil {
-		return exporterhelper.NumTimeSeries(md), consumererror.Permanent(err)
+		return numTimeseries, consumererror.Permanent(err)
 	}
 
 	req, err := http.NewRequest("POST", s.ingestURL.String(), body)
 	if err != nil {
-		return exporterhelper.NumTimeSeries(md), consumererror.Permanent(err)
+		return numTimeseries, consumererror.Permanent(err)
 	}
 
-	for k, v := range s.headers {
-		req.Header.Set(k, v)
+	// Override access token in headers map if it's non empty.
```

Comment:
I think case sensitivity issues are introduced here that were addressed by using the `req.Header` interface directly.

Reply:
It looks like we previously set the header using `req.Header.Set(splunk.SFxAccessTokenHeader, accessToken)`, if `accessToken` is non-empty. Wouldn't we have had the same issue before this update?

Reply:
since `headers` is is map, if for whatever reason a user configured an `"x-sf-token"` entry and we set ` "X-Sf-Token"` here, the subsequent `req.Header.Set(k,v)` call will be nondeterministic since it's case insensitive.  By using `Header.Set` after setting the headers individually we're guaranteed to overwrite: https://golang.org/pkg/net/http/#Header.Set

Reply:
Ah I see. Thanks for the clarification. Will update this.

Reply:
Updated with a fix for this issue and also added a test around this.